### DISCUSSION TITLE
`mandatoryHours` greifen nicht

### DISCUSSION THREAD

**Discussion by deadrabbit87** [camueller/node-red-contrib-heatpump-tibber-pv#7] [Allgemein]
Ich bin jetzt schon seit mehreren Wochen an dem Flow dran, aber irgendwie macht er nicht das, wie ich es mir vorstelle. 

Hierbei sind mir mehrere Probleme aufgefallen. 

Die `mandatoryHours` greifen bei mir irgendwie nicht. Gesetzt sind diese so `["0-6","12-17",2]`. 
Was ich so verstehe, dass die Wärmepumpe zwingend jeden Wochentag (`0-6` = Sonntag - Montag) zwischen 12 und 17 Uhr mindestens zwei Stunden laufen muss. Da ich `powerConsumptionHours` auf 8 gesetzt habe, würde ich das so verstehen, dass von den 8 Stunden die zwei Stunden abgezogen werden. Übrig bleiben sollten dann 6 Stunden, die nach PV und Strompreis verteilt werden. 

Was aber rauskommt ist das: 
![grafik](https://github.com/user-attachments/assets/f204c50e-1493-408c-ade4-f78ee4cd51ee)

Merkwürdigerweise ist das auch jeden Tag das gleiche. Es ändert sich nicht, obwohl die API-Calls funktionieren und sich die PV-Prognose als auch der Strompreis- und Kostengraph ändert. 

`fallbackHours` sehen so aus` [["0-5"],[6,8],[0,8]]`. Hätte ich jetzt als Fehlerursache mal ausgeschlossen.

Wo setze ich da am Besten zum debuggen an?

**Comment by camueller** [comment]
Bei Dir werden mehr als 8 Stunden als "kostengünstigste Stunden" identifiziert. Da stimmt vermutlich etwas mit der Konfiguration nicht.

Funktioniert denn der Flow ohne mandatoryHours und fallbsckHours?

Verwendest Du die aktuelle Version des Flows?

**Reply by deadrabbit87** [reply]
Ja, entferne ich `mandatoryHours` funktioniert der Flow (`fallbackHours` ist noch drin): 
 
![grafik](https://github.com/user-attachments/assets/09ed4815-3f5b-44f8-975c-b49432e558fd)

Die Version ist die aktuelle aus dem Repo. Ich hab diesen nur hinsichtlich des GPIOs angepasst, da ich einen Shelly am SG Ready Kontakt nutze. Also der Teil nach `is selected hour?`

**Comment by camueller** [comment]
Habe gerade nochmal bei mir selbst nachgesehen: Ich denke, dass es daran liegt, dass Du die äusseren Klammern vergessen hast. Es sollte so aussehen:

[["0-6","12-17",2]]

**Reply by deadrabbit87** [reply]
Danke - ich glaube das wars. Ich hätte zwar schwören können, dass ich das zuerst so hatte und es auch nicht funktioniert hat, aber was solls...

**Comment by deadrabbit87** [comment]
Um das hier nochmal aufzuwärmen: Zwei `mandatoryHours` gehen wohl nicht oder?
Also so: 
`[["0-6","12-17",3],["0-6","1-7",2]]`

Danke!

**Comment by camueller** [comment]
Warum soll das nicht gehen? Bei mir sieht das so aus:
```
[["1-5",4],[6,"0-8",4,[8]],[0,"0-8",4,[8]]]
```

**Reply by deadrabbit87** [reply]
Um das hier nochmal aufzuwärmen: 
Setzt man 2 x 3 Elemente, geht das nicht: 
`[["0-6","1-6",2],["0-6","12-17",3]]`

Und das Ergebnis: 
![grafik](https://github.com/user-attachments/assets/4c23ef71-d228-4407-a8e0-7630a419e3ec)

Nach meinem Vertändnis sollte hier die WP min. 3 h in der Zeit von 12-17 Uhr laufen. Tut Sie aber nicht. 

Das hier wäre auch der JSON String aus choose cost efficient hours: 
```json
["2025-01-30T00","2025-01-30T01","2025-01-30T02","2025-01-30T03","2025-01-30T04","2025-01-30T05","2025-01-30T12","2025-01-30T13"]
```

**Reply by camueller** [reply]
Ich muss mal schauen, ob das Problem auch bei `node-red-contrib-heatpump-dynamic-tariff`besteht. Dann würde ich das dort fixen. Für `node-red-contrib-heatpump-tibber-pv`wird es keine Fixes mehr geben.

**Reply by deadrabbit87** [reply]
Ja klar, ich kann da auch ein Issue aufmachen. 

Wenn liegt das Problem eh in dem Javascript von dem Node.

**Reply by camueller** [reply]
Habe den Fehler gefunden. Du musst drei Zeilen anpassen im Knoten "choose cost efficient hours":

![image](https://github.com/user-attachments/assets/92dfb2a0-7db1-4d08-9066-b1a344d6dbe0)

**Reply by deadrabbit87** [reply]
Perfekt - danke, damit gehts dann.
